File: carebridge-mobile/carebridge_theme/lib/themes/app_fonts.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'app_colors.dart';

class AppFonts {
  final double? fontSize;
  final FontWeight? fontWeight;
  final Color? color;
  final TextDecoration? decoration;
  final FontStyle? fontStyle;

  AppFonts._({
    this.fontSize,
    this.fontWeight,
    this.color,
    this.decoration,
    this.fontStyle,
  });

  AppFonts copyWith({
    double? fontSize,
    FontWeight? fontWeight,
    Color? color,
    TextDecoration? decoration,
    FontStyle? fontStyle,
  }) {
    return AppFonts._(
      fontSize: fontSize ?? this.fontSize,
      fontWeight: fontWeight ?? this.fontWeight,
      color: color ?? this.color,
      decoration: decoration ?? this.decoration,
      fontStyle: fontStyle ?? this.fontStyle,
    );
  }

  TextStyle get ts {
    return GoogleFonts.inter(
      fontSize: fontSize ?? _body.fontSize,
      fontWeight: fontWeight ?? FontWeight.normal,
      color: color ?? text.color,
      letterSpacing: 0,
      decoration: decoration,
      fontStyle: fontStyle,
    );
  }

  /// Color
  AppFonts customColor(Color color) {
    return copyWith(color: color);
  }

  AppFonts get placeholder {
    return copyWith(color: appColors.placeholder);
  }

  AppFonts get text {
    return copyWith(color: appColors.font);
  }

  AppFonts get light {
    return copyWith(color: appColors.lightFont);
  }

  AppFonts get gray {
    return copyWith(color: appColors.grayFont);
  }

  AppFonts get disabled {
    return copyWith(color: appColors.disabledButtonFont);
  }

  AppFonts get primary {
    return copyWith(color: appColors.primary);
  }

  AppFonts get secondary {
    return copyWith(color: appColors.secondary);
  }

  AppFonts get white {
    return copyWith(color: appColors.white);
  }

  AppFonts get info {
    return copyWith(color: appColors.info);
  }

  AppFonts get success {
    return copyWith(color: appColors.success);
  }

  AppFonts get warning {
    return copyWith(color: appColors.warning);
  }

  AppFonts get error {
    return copyWith(color: appColors.error);
  }

  /// Decoration
  AppFonts customWeight(FontWeight fontWeight) {
    return copyWith(fontWeight: fontWeight);
  }

  AppFonts get bold {
    return copyWith(fontWeight: FontWeight.bold);
  }

  AppFonts get semibold {
    return copyWith(fontWeight: FontWeight.w600);
  }

  AppFonts get underline {
    return copyWith(decoration: TextDecoration.underline);
  }

  AppFonts get italic {
    return copyWith(fontStyle: FontStyle.italic);
  }

  /// Font Size
  AppFonts customSize(double fontSize) {
    return copyWith(fontSize: fontSize);
  }

  AppFonts get captionSmall {
    return copyWith(fontSize: 8);
  }

  AppFonts get caption {
    return copyWith(fontSize: 12);
  }

  AppFonts get _body {
    return copyWith(fontSize: 14);
  }

  AppFonts get subtitle {
    return copyWith(fontSize: 18);
  }

  AppFonts get titleSmall {
    return copyWith(fontSize: 24);
  }

  AppFonts get title {
    return copyWith(fontSize: 32);
  }
}

///
/// Example: appFonts.title.ts
///

AppFonts get appFonts {
  return AppFonts._();
}

File: carebridge-admin-nextjs/src/themes/app_fonts.ts
import { appColors } from './app_colors';

interface TextStyle {
    fontSize?: number;
    fontWeight?: string;
    color?: string;
    letterSpacing?: number;
    decoration?: 'underline' | 'line-through' | 'none';
    fontStyle?: 'normal' | 'italic' | 'oblique';
}

class AppFonts {
    readonly fontSize?: number;
    readonly fontWeight?: string;
    readonly color?: string;
    readonly decoration?: 'underline' | 'line-through' | 'none';
    readonly fontStyle?: 'normal' | 'italic' | 'oblique';

    private constructor(
        fontSize?: number,
        fontWeight?: string,
        color?: string,
        decoration?: 'underline' | 'line-through' | 'none',
        fontStyle?: 'normal' | 'italic' | 'oblique'
    ) {
        this.fontSize = fontSize;
        this.fontWeight = fontWeight;
        this.color = color;
        this.decoration = decoration;
        this.fontStyle = fontStyle;
    }

    static create(
        fontSize?: number,
        fontWeight?: string,
        color?: string,
        decoration?: 'underline' | 'line-through' | 'none',
        fontStyle?: 'normal' | 'italic' | 'oblique'
    ): AppFonts {
        return new AppFonts(fontSize, fontWeight, color, decoration, fontStyle);
    }

    copyWith(updateValues: Partial<AppFonts>): AppFonts {
        return new AppFonts(
            updateValues.fontSize ?? this.fontSize,
            updateValues.fontWeight ?? this.fontWeight,
            updateValues.color ?? this.color,
            updateValues.decoration ?? this.decoration,
            updateValues.fontStyle ?? this.fontStyle
        );
    }

    get ts(): TextStyle {
        return {
            fontSize: this.fontSize ?? 14,
            fontWeight: this.fontWeight ?? 'normal',
            color: this.color ?? appColors.neutral[80],
            letterSpacing: 0,
            decoration: this.decoration ?? 'none',
            fontStyle: this.fontStyle ?? 'normal',
        };
    }

    // Color Methods
    customColor(color: string): AppFonts {
        return this.copyWith({ color });
    }

    get placeholder(): AppFonts {
        return this.copyWith({ color: appColors.neutral[50] });
    }

    get text(): AppFonts {
        return this.copyWith({ color: appColors.neutral[80] });
    }

    get light(): AppFonts {
        return this.copyWith({ color: appColors.neutral[60] });
    }

    get gray(): AppFonts {
        return this.copyWith({ color: appColors.neutral[70] });
    }

    get disabled(): AppFonts {
        return this.copyWith({ color: appColors.neutral[40] });
    }

    get primary(): AppFonts {
        return this.copyWith({ color: appColors.primary[500] });
    }

    get secondary(): AppFonts {
        return this.copyWith({ color: appColors.secondary[500] });
    }

    get white(): AppFonts {
        return this.copyWith({ color: appColors.neutral[0] });
    }

    get info(): AppFonts {
        return this.copyWith({ color: appColors.info[20] });
    }

    get success(): AppFonts {
        return this.copyWith({ color: appColors.success[20] });
    }

    get warning(): AppFonts {
        return this.copyWith({ color: appColors.warning[20] });
    }

    get error(): AppFonts {
        return this.copyWith({ color: appColors.error[20] });
    }

    // Decoration Methods
    customWeight(fontWeight: string): AppFonts {
        return this.copyWith({ fontWeight });
    }

    get bold(): AppFonts {
        return this.copyWith({ fontWeight: 'bold' });
    }

    get semibold(): AppFonts {
        return this.copyWith({ fontWeight: '600' });
    }

    get underline(): AppFonts {
        return this.copyWith({ decoration: 'underline' });
    }

    get italic(): AppFonts {
        return this.copyWith({ fontStyle: 'italic' });
    }

    // Font Size Methods
    customSize(fontSize: number): AppFonts {
        return this.copyWith({ fontSize });
    }

    get captionSmall(): AppFonts {
        return this.copyWith({ fontSize: 8 });
    }

    get caption(): AppFonts {
        return this.copyWith({ fontSize: 12 });
    }

    get body(): AppFonts {
        return this.copyWith({ fontSize: 14 });
    }

    get subtitle(): AppFonts {
        return this.copyWith({ fontSize: 18 });
    }

    get titleSmall(): AppFonts {
        return this.copyWith({ fontSize: 24 });
    }

    get title(): AppFonts {
        return this.copyWith({ fontSize: 32 });
    }
}

export const appFonts = AppFonts.create();


// example usage
// const styles = appFonts
//     .customSize(16)
//     .customWeight('bold')
//     .customColor(AppColors.primary[500])
//     .customDecoration('underline')
//     .customStyle('italic')
//     .ts;

// console.log(styles); // { fontSize: 16, fontWeight: 'bold', color: 'var(--color-primary-500)', letterSpacing: 0, decoration: 'underline', fontStyle: 'italic' }

// const styles2 = appFonts
//     .body
//     .bold
//     .primary
//     .underline
//     .italic